spidermonkey: Add riscv64 support

zhaofengli opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This PR adds riscv64 support to spidermonkey78. The included patches are already merged upstream, and a modified version that applies on 78 is included in-tree.

Tested by building natively on riscv64 and x86-64.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

andir commented on pkgs/development/interpreters/spidermonkey/78.nix:
```diff
@@ -27,6 +27,18 @@ stdenv.mkDerivation rec {
   };
 
+  patches = []
+    ++ lib.optionals stdenv.hostPlatform.isRiscV [
```
Can we apply the patch unconditionally? If it isn't required to be conditional we shouldn't as otherwise it might just "rot" without anyone (that is using `riscv`) noticing.

zhaofengli replied:
Ok, I've made the patch unconditional and rebased to staging.

zhaofengli committed:
spidermonkey: Force lp64d ABI for riscv64

cc-rs is hardcoded to use the soft-float ABI for riscv64, while
the rest of the system is double-float. This is not just a missing
stub header file - You can't link object files using different
ABIs.

zhaofengli committed:
spidermonkey: Add patch to support riscv64

ius reviewed (approved):
I have no meaningful way to test this, but it builds fine for RISC-V (cross) and review suggests it should be a no-op for other arches.

Can we rebase and merge this?

zhaofengli commented:
> I have no meaningful way to test this, but it builds fine for RISC-V (cross) and review suggests it should be a no-op for other arches.

You can run the `js` executable under binfmt and it should give you a JS REPL.

> Can we rebase and merge this?

Sure, rebased on staging and fixed the conflict.

ius commented:
> You can run the `js` executable under binfmt and it should give you a JS REPL.

Ah! Didn't know it ships a REPL.

In that case:

```
$ qemu-riscv64 /nix/store/685c24yaqfliprb2cixjc5rlb8f5rh7v-spidermonkey-riscv64-unknown-linux-gnu-78.15.0/bin/js78 
js> 0.42 + 13
13.42
```

Works for me.

lostnet reviewed (approved):
Looks fine for existing archs to me.
